File: app/src/main/java/org/tasks/widget/AppWidgetManager.java
package org.tasks.widget;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import javax.inject.Inject;
import org.tasks.R;
import org.tasks.injection.ApplicationContext;

public class AppWidgetManager {

  private final android.appwidget.AppWidgetManager appWidgetManager;
  private final Context context;

  @Inject
  public AppWidgetManager(@ApplicationContext Context context) {
    this.context = context;
    appWidgetManager = android.appwidget.AppWidgetManager.getInstance(context);
  }

  public void updateWidgets() {
    appWidgetManager.notifyAppWidgetViewDataChanged(getWidgetIds(), R.id.list_view);
  }

  public void reconfigureWidgets(int... appWidgetIds) {
    Intent intent = new Intent(context, TasksWidget.class);
    intent.setAction(android.appwidget.AppWidgetManager.ACTION_APPWIDGET_UPDATE);
    intent.putExtra(
        android.appwidget.AppWidgetManager.EXTRA_APPWIDGET_IDS,
        appWidgetIds.length == 0 ? getWidgetIds() : appWidgetIds);
    context.sendBroadcast(intent);
    updateWidgets();
  }

  public int[] getWidgetIds() {
    return appWidgetManager.getAppWidgetIds(new ComponentName(context, TasksWidget.class));
  }
}

File: app/src/main/java/org/tasks/widget/AppWidgetManager.kt
package org.tasks.widget

import android.appwidget.AppWidgetManager
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import dagger.hilt.android.qualifiers.ApplicationContext
import timber.log.Timber
import javax.inject.Inject

class AppWidgetManager @Inject constructor(
    @param:ApplicationContext private val context: Context,
) {
    private val appWidgetManager: AppWidgetManager? by lazy {
        AppWidgetManager.getInstance(context)
    }

    val widgetIds: IntArray
        get() = appWidgetManager
                ?.getAppWidgetIds(ComponentName(context, TasksWidget::class.java))
                ?: intArrayOf()

    fun rebuildWidgets(vararg appWidgetIds: Int) {
        if (appWidgetIds.isEmpty()) {
            return
        }
        Timber.d("rebuildWidgets(${appWidgetIds.joinToString()})")
        val intent = Intent(context, TasksWidget::class.java)
            .putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, appWidgetIds)
            .apply { action = AppWidgetManager.ACTION_APPWIDGET_UPDATE }
        context.sendBroadcast(intent)
    }

    fun updateWidgets() {
        rebuildWidgets(*widgetIds)
    }

    fun exists(id: Int) = appWidgetManager?.getAppWidgetInfo(id) != null
}
